File: PhoneticTokenizer/go/phonetok/lexicon.go
package phonetok

import (
	"bufio"
	_ "embed"
	"fmt"
	"io"
	"os"
	"sort"
	"strings"
)

// The pronouncing lexicon (the port of lexicon.py): CMU-format entries, layered
// from the bundled core and the file named by PHONETOK_LEXICON.

//go:embed data/core.dict
var coreDict string

// EnvLexicon names the environment variable that points at a full dictionary file.
const EnvLexicon = "PHONETOK_LEXICON"

// Entry is one (word, pronunciation) pair.
type Entry struct {
	Word   string
	Phones []string
}

// ParseEntry reads one line of a CMU-format file; ok is false for a comment or a blank.
func ParseEntry(line string) (word string, phones []string, ok bool) {
	if strings.HasPrefix(line, ";;;") {
		return "", nil, false
	}
	if i := strings.IndexByte(line, '#'); i >= 0 {
		line = line[:i]
	}
	parts := strings.Fields(line)
	if len(parts) < 2 {
		return "", nil, false
	}
	word = strings.ToLower(parts[0])
	if strings.HasSuffix(word, ")") {
		if i := strings.LastIndexByte(word, '('); i >= 0 {
			word = word[:i]
		}
	}
	phones = make([]string, len(parts)-1)
	for i, p := range parts[1:] {
		phones[i] = strings.ToUpper(p)
	}
	return word, phones, true
}

// ReadEntries parses every entry of a CMU-format stream.
func ReadEntries(r io.Reader) ([]Entry, error) {
	var out []Entry
	sc := bufio.NewScanner(r)
	sc.Buffer(make([]byte, 1<<20), 1<<20)
	for sc.Scan() {
		if w, p, ok := ParseEntry(sc.Text()); ok {
			out = append(out, Entry{w, p})
		}
	}
	return out, sc.Err()
}

// CoreEntries are the bundled core lexicon's entries.
func CoreEntries() []Entry {
	entries, _ := ReadEntries(strings.NewReader(coreDict))
	return entries
}

// Lexicon is a word -> pronunciations map and its inverse.  Every method
// lower-cases the word it is given.
type Lexicon struct {
	entries map[string][][]string
	order   []string // words in the order first added, so that listings are stable
	reverse map[string][]string
	loose   map[string][]string
	Sources []string
}

// NewLexicon is an empty lexicon.
func NewLexicon() *Lexicon {
	return &Lexicon{entries: map[string][][]string{}}
}

// CoreLexicon is the bundled core lexicon.
func CoreLexicon() *Lexicon {
	lex := NewLexicon()
	lex.Extend(CoreEntries())
	lex.Sources = append(lex.Sources, "the core lexicon")
	return lex
}

// LoadLexicon reads a CMU-format file.
func LoadLexicon(path string) (*Lexicon, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	entries, err := ReadEntries(f)
	if err != nil {
		return nil, err
	}
	lex := NewLexicon()
	lex.Extend(entries)
	lex.Sources = append(lex.Sources, "file "+path)
	return lex, nil
}

// PortableLexicon is the lexicon every port builds alike: the core, plus the
// PHONETOK_LEXICON file if it is set (the Python package's Lexicon.portable()).
func PortableLexicon() (*Lexicon, error) {
	lex := CoreLexicon()
	if path := os.Getenv(EnvLexicon); path != "" {
		f, err := os.Open(path)
		if err != nil {
			return nil, err
		}
		defer f.Close()
		entries, err := ReadEntries(f)
		if err != nil {
			return nil, err
		}
		lex.Extend(entries)
		lex.Sources = append(lex.Sources, "file "+path)
	}
	return lex, nil
}

func samePhones(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

// Extend adds entries; a pronunciation the word already has is not added twice.  Returns how many were new.
func (l *Lexicon) Extend(entries []Entry) int {
	added := 0
	for _, e := range entries {
		word := strings.ToLower(e.Word)
		phones := append([]string(nil), e.Phones...)
		known, ok := l.entries[word]
		if !ok {
			l.entries[word] = [][]string{phones}
			l.order = append(l.order, word)
			added++
			continue
		}
		dup := false
		for _, k := range known {
			if samePhones(k, phones) {
				dup = true
				break
			}
		}
		if !dup {
			l.entries[word] = append(known, phones)
			added++
		}
	}
	if added > 0 {
		l.reverse, l.loose = nil, nil
	}
	return added
}

// Add adds a pronunciation; first makes it the one Lookup returns.
func (l *Lexicon) Add(word string, phones []string, first bool) error {
	if err := CheckPhones(phones); err != nil {
		return err
	}
	word = strings.ToLower(word)
	phones = append([]string(nil), phones...)
	known := l.entries[word]
	if known == nil {
		l.order = append(l.order, word)
	}
	kept := known[:0:0]
	for _, k := range known {
		if !samePhones(k, phones) {
			kept = append(kept, k)
		}
	}
	if first {
		kept = append([][]string{phones}, kept...)
	} else {
		kept = append(kept, phones)
	}
	l.entries[word] = kept
	l.reverse, l.loose = nil, nil
	return nil
}

// Len is how many words the lexicon holds.
func (l *Lexicon) Len() int { return len(l.entries) }

// Has reports whether a word is in the lexicon.
func (l *Lexicon) Has(word string) bool {
	_, ok := l.entries[strings.ToLower(word)]
	return ok
}

// Lookup is the first pronunciation of a word, or nil.
func (l *Lexicon) Lookup(word string) []string {
	variants := l.entries[strings.ToLower(word)]
	if len(variants) == 0 {
		return nil
	}
	return variants[0]
}

// Pronunciations is every pronunciation of a word; nil if it is unknown.
func (l *Lexicon) Pronunciations(word string) [][]string {
	return l.entries[strings.ToLower(word)]
}

// Words are the words in the order they were first added.
func (l *Lexicon) Words() []string { return l.order }

// Items is every (word, pronunciation) pair, alternatives included, in first-added order.
func (l *Lexicon) Items() []Entry {
	var out []Entry
	for _, w := range l.order {
		for _, p := range l.entries[w] {
			out = append(out, Entry{w, p})
		}
	}
	return out
}

func (l *Lexicon) buildReverse() {
	l.reverse, l.loose = map[string][]string{}, map[string][]string{}
	for _, w := range l.order {
		for _, p := range l.entries[w] {
			key := strings.Join(p, " ")
			l.reverse[key] = append(l.reverse[key], w)
			loose := strings.Join(StripStress(p), " ")
			l.loose[loose] = append(l.loose[loose], w)
		}
	}
}

// Spellings are the words pronounced exactly like phones (homophones), in the order they were added;
// with stress false the stress digits are ignored on both sides.
func (l *Lexicon) Spellings(phones []string, stress bool) []string {
	if l.reverse == nil {
		l.buildReverse()
	}
	if stress {
		return append([]string(nil), l.reverse[strings.Join(phones, " ")]...)
	}
	return append([]string(nil), l.loose[strings.Join(StripStress(phones), " ")]...)
}

// Describe is "N words from <sources>".
func (l *Lexicon) Describe() string {
	src := "nowhere"
	if len(l.Sources) > 0 {
		src = strings.Join(l.Sources, ", ")
	}
	return fmt.Sprintf("%d words from %s", len(l.entries), src)
}

// SortedWords are the words in alphabetical order (for listings).
func (l *Lexicon) SortedWords() []string {
	out := append([]string(nil), l.order...)
	sort.Strings(out)
	return out
}

File: PhoneticTokenizer/phonetok/lexicon.py
"""The pronouncing lexicon: which sounds a spelled word is made of, and which words a sound sequence spells.

The format is the CMU Pronouncing Dictionary's, one entry per line::

    word  PH ON ES
    word(2)  PH ON ES        an alternative pronunciation
    ;;; a comment            (a trailing ``# comment`` is dropped too)

Three sources, layered:

* the **core** lexicon bundled with the package (``data/core.dict``: the
  common words, the numbers, the letters - some 1 900 entries, always there);
* the **full** dictionary, when it can be found: a file named by the
  ``PHONETOK_LEXICON`` environment variable, or the ``cmudict`` package
  (``pip install cmudict``: 135 000 words);
* whatever the caller :meth:`Lexicon.add`\\ s.

A word the lexicon does not hold is sounded out by :mod:`phonetok.rules`.
"""

from __future__ import annotations

import importlib.util
import os
from collections.abc import Iterable, Iterator
from typing import IO

from .phones import check_phones, strip_stress

CORE_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data", "core.dict")
ENV_LEXICON = "PHONETOK_LEXICON"

Phones = tuple[str, ...]


def parse_entry(line: str) -> tuple[str, Phones] | None:
    """One line of a CMU-format file as ``(word, phones)``; ``None`` for a comment or a blank."""
    if line.startswith(";;;"):
        return None
    line = line.split("#", 1)[0].strip()
    if not line:
        return None
    parts = line.split()
    if len(parts) < 2:
        return None
    word = parts[0].lower()
    if word.endswith(")") and "(" in word:  # an alternative: cat(2)
        word = word[:word.rindex("(")]
    return word, tuple(p.upper() for p in parts[1:])


def read_entries(lines: Iterable[str]) -> Iterator[tuple[str, Phones]]:
    for line in lines:
        entry = parse_entry(line)
        if entry is not None:
            yield entry


_CORE_CACHE: list[tuple[str, Phones]] | None = None


def core_entries() -> list[tuple[str, Phones]]:
    """The bundled core lexicon's entries, parsed once per process."""
    global _CORE_CACHE
    if _CORE_CACHE is None:
        with open(CORE_PATH, encoding="utf-8") as f:
            _CORE_CACHE = list(read_entries(f))
    return list(_CORE_CACHE)


def find_full_dictionary(package: bool = True) -> tuple[str, IO[str] | None] | None:
    """Where the full dictionary is, if anywhere: ``(description, open file)``.

    The ``PHONETOK_LEXICON`` variable names a file; failing that the ``cmudict``
    package is used if it is installed (and ``package`` is true).  ``None`` when
    neither is there.
    """
    path = os.environ.get(ENV_LEXICON)
    if path:
        return f"file {path}", open(path, encoding="utf-8")
    if package and importlib.util.find_spec("cmudict") is not None:
        import cmudict  # type: ignore

        return "the cmudict package", cmudict.dict_stream()
    return None


class Lexicon:
    """A word -> pronunciations map, and its inverse.

    ``Lexicon()`` is empty; :meth:`core` is the bundled subset; :meth:`default`
    is the core plus the full dictionary when one can be found.  Every method
    lower-cases the word it is given, since the dictionary is lower case.
    """

    __slots__ = ("_entries", "_reverse", "_reverse_loose", "sources")

    def __init__(self, entries: Iterable[tuple[str, Iterable[str]]] = (), source: str | None = None) -> None:
        self._entries: dict[str, list[Phones]] = {}
        self._reverse: dict[Phones, list[str]] | None = None
        self._reverse_loose: dict[Phones, list[str]] | None = None
        self.sources: list[str] = []
        self.extend(entries)
        if source:
            self.sources.append(source)

    # -- construction --------------------------------------------------------

    @classmethod
    def core(cls) -> "Lexicon":
        """The bundled core lexicon."""
        return cls(core_entries(), source="the core lexicon")

    @classmethod
    def load(cls, path: str) -> "Lexicon":
        """A CMU-format file."""
        with open(path, encoding="utf-8") as f:
            return cls(read_entries(f), source=f"file {path}")

    @classmethod
    def default(cls, full: bool = True, package: bool = True) -> "Lexicon":
        """The core lexicon, plus the full dictionary if ``full`` and one can be found.

        ``package=False`` ignores the ``cmudict`` package and reads only the file named by
        ``PHONETOK_LEXICON``: the two sources the Go and Rust ports can read too.
        """
        lex = cls.core()
        if full:
            found = find_full_dictionary(package=package)
            if found is not None:
                name, stream = found
                try:
                    lex.extend(read_entries(stream))
                finally:
                    if stream is not None:
                        stream.close()
                lex.sources.append(name)
        return lex

    @classmethod
    def portable(cls) -> "Lexicon":
        """The lexicon every port builds alike: the core, plus the ``PHONETOK_LEXICON`` file if it is set.

        A model whose symbols are sounds is only as portable as the lexicon that made them, so
        a Python model meant for Go or Rust is transcribed through this one.
        """
        return cls.default(full=True, package=False)

    def extend(self, entries: Iterable[tuple[str, Iterable[str]]]) -> int:
        """Add entries; a pronunciation the word already has is not added twice.  Returns how many were new."""
        added = 0
        table = self._entries
        for word, phones in entries:
            phones = tuple(phones)
            word = word.lower()
            known = table.get(word)
            if known is None:
                table[word] = [phones]
                added += 1
            elif phones not in known:
                known.append(phones)
                added += 1
        if added:
            self._reverse = self._reverse_loose = None
        return added

    def add(self, word: str, phones: Iterable[str], first: bool = False) -> None:
        """Add a pronunciation; ``first`` makes it the one :meth:`lookup` returns."""
        phones = tuple(phones)
        check_phones(phones)
        word = word.lower()
        known = self._entries.setdefault(word, [])
        if phones in known:
            known.remove(phones)
        if first:
            known.insert(0, phones)
        else:
            known.append(phones)
        self._reverse = self._reverse_loose = None

    def remove(self, word: str) -> bool:
        gone = self._entries.pop(word.lower(), None) is not None
        if gone:
            self._reverse = self._reverse_loose = None
        return gone

    # -- lookup --------------------------------------------------------------

    def __len__(self) -> int:
        return len(self._entries)

    def __contains__(self, word: object) -> bool:
        return isinstance(word, str) and word.lower() in self._entries

    def words(self) -> Iterator[str]:
        return iter(self._entries)

    def items(self) -> Iterator[tuple[str, Phones]]:
        """Every ``(word, pronunciation)`` pair, alternatives included."""
        for word, variants in self._entries.items():
            for phones in variants:
                yield word, phones

    def lookup(self, word: str) -> Phones | None:
        """The first pronunciation of ``word``, or ``None``."""
        variants = self._entries.get(word.lower())
        return variants[0] if variants else None

    def pronunciations(self, word: str) -> list[Phones]:
        """Every pronunciation of ``word``; ``[]`` if it is unknown."""
        return list(self._entries.get(word.lower(), ()))

    # -- the inverse ---------------------------------------------------------

    def _build_reverse(self) -> None:
        exact: dict[Phones, list[str]] = {}
        loose: dict[Phones, list[str]] = {}
        for word, variants in self._entries.items():
            for phones in variants:
                exact.setdefault(phones, []).append(word)
                loose.setdefault(tuple(strip_stress(phones)), []).append(word)
        self._reverse, self._reverse_loose = exact, loose

    def spellings(self, phones: Iterable[str], stress: bool = True) -> list[str]:
        """The words pronounced exactly like ``phones`` (homophones), in the order they were added.

        With ``stress=False`` the stress digits are ignored on both sides.
        """
        if self._reverse is None:
            self._build_reverse()
        assert self._reverse is not None and self._reverse_loose is not None
        phones = tuple(phones)
        if stress:
            return list(self._reverse.get(phones, ()))
        return list(self._reverse_loose.get(tuple(strip_stress(phones)), ()))

    def describe(self) -> str:
        return f"{len(self)} words from " + (", ".join(self.sources) if self.sources else "nowhere")

    def __repr__(self) -> str:
        return f"Lexicon({self.describe()})"
